antsibull/write_docs.py

```diff
@@ -218,6 +245,48 @@ async def output_collection_index(collection_info: CollectionInfoT,
     flog.debug('Leave')
 
 
+async def output_plugin_indexes(collection_info: CollectionInfoT,
+                                dest_dir: str) -> None:
+    """
+    Generate top-level plugin index pages for all plugins of a type in all collections.
+
+    :arg collection_info: Mapping of collection_name to Mapping of plugin_type to Mapping of
+        collection_name to short_description.
+    :arg dest_dir: The directory to place the documentation in.
+    """
+    flog = mlog.fields(func='output_plugin_indexes')
+    flog.debug('Enter')
+
+    env = doc_environment(('antsibull.data', 'docsite'))
+    # Get the templates
+    plugin_list_tmpl = env.get_template('list_of_plugins.rst.j2')
+
+    collection_toplevel = os.path.join(dest_dir, 'collections')
+    flog.fields(toplevel=collection_toplevel, exists=os.path.isdir(collection_toplevel)).debug(
+        'collection_toplevel exists?')
+    # This is only safe because we made sure that the top of the directory tree we're writing to
+    # (docs/docsite/rst) is only writable by us.
+    os.makedirs(collection_toplevel, mode=0o755, exist_ok=True)
+
+    plugin_types = set()
+    for collection_name, plugin_maps in collection_info.items():
+        for plugin_type in plugin_maps:
+            plugin_types.add(plugin_type)
+
+    writers = []
+    lib_ctx = app_context.lib_ctx.get()
+    async with asyncio_pool.AioPool(size=lib_ctx.thread_max) as pool:
+        for plugin_type in plugin_types:
+            filename = os.path.join(
+                collection_toplevel, 'index_{type}.rst'.format(type=plugin_type))
+            writers.append(await pool.spawn(
+                write_plugin_type_index(plugin_type, collection_info, plugin_list_tmpl, filename)))
```

Comment:
Heh.  So I looked at this and started writing some code and then looked further up the stack and realized that we probably have everything we need already.  Sorta.  Almost.  Here's my thought:

* I think we want the code in this file to be about formatting and then writing output.  Partitioning the data seems like it is okay as that is part of formatting (so splitting the output up into a page for each plugin type is okay) but the data we get shouldn't need to be restructured for that inside of this file. 
* So the input data should be restructured in the caller which is `generate_docs_for_all_collections()`in the cli code in `stable.py`.
* Looking at the code in `generate_docs_for_all_collections()`, I see that we almost have the data in the format we want in the `plugin_info` variable.  That currently holds the raw data from the plugins; a little more work is needed to handle errors that were listed in the raw data.
* `get_collection_contents()` handles the errors and reformats the fqcn to short names with collection as key but it also restructures the data to be mapped by collection instead of plugin.

* So I think what we need to do is separate `get_collection_contents()` into two functions.  First function would be something like `plugin_contents = get_plugin_contents(plugin_info)` which would handle the errors and transforming fqcn.  Then we can call a revised `collection_info = get_collection_contents(plugin_contents)` which only restructures that data.  At that point we'll have two variables, `plugin_contents` which we can pass in here and `collection_info` which we can pass into `output_collection_index`.

Does all that make sense?

Reply:
I think I applied everything you wrote in d0cdb0a8e2cef539432d7798e189f6b88648a68a. It produces exactly the same result as before.